File: src/main/java/com/optum/ghs/grommet/Scanner.java
package com.optum.ghs.grommet;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import com.optum.ghs.grommet.types.Token;
import com.optum.ghs.grommet.types.TokenType;

// characters to tokens
public class Scanner {
    private static final Map<String, TokenType> keywords;
    static {
        keywords = new HashMap<>();
        keywords.put("and", TokenType.AND);
        keywords.put("class", TokenType.CLASS);
        keywords.put("else", TokenType.ELSE);
        keywords.put("false", TokenType.FALSE);
        keywords.put("for", TokenType.FOR);
        keywords.put("fun", TokenType.FUN);
        keywords.put("if", TokenType.IF);
        keywords.put("nil", TokenType.NIL);
        keywords.put("or", TokenType.OR);
        keywords.put("print", TokenType.PRINT);
        keywords.put("return", TokenType.RETURN);
        keywords.put("super", TokenType.SUPER);
        keywords.put("this", TokenType.THIS);
        keywords.put("true", TokenType.TRUE);
        keywords.put("var", TokenType.VAR);
        keywords.put("while", TokenType.WHILE);
    }
    private final String source;
    private final List<Token> tokens = new ArrayList<>();
    private int start = 0;
    private int current = 0;
    private int line = 1;

    Scanner(String source) {
        this.source = source;
    }

    List<Token> scanTokens() {
        while (!isAtEnd()) {
            // start at the beginning of the next lexeme
            this.start = this.current;
            scanToken();
        }

        this.tokens.add(new Token(TokenType.EOF, "", null, line)); // add eof at the end
        return this.tokens;
    }

    // scan a single token
    void scanToken() {
        char c = advance();
        switch (c) {
            case '(':
                addToken(TokenType.LEFT_PAREN);
                break;
            case ')':
                addToken(TokenType.RIGHT_PAREN);
                break;
            case '{':
                addToken(TokenType.LEFT_BRACE);
                break;
            case '}':
                addToken(TokenType.RIGHT_BRACE);
                break;
            case ',':
                addToken(TokenType.COMMA);
                break;
            case '.':
                addToken(TokenType.DOT);
                break;
            case '-':
                addToken(match('=') ? TokenType.MINUSEQUALS : TokenType.MINUS);
                break;
            case '+':
                addToken(match('=') ? TokenType.PLUSEQUALS : TokenType.PLUS);
                break;
            case ';':
                addToken(TokenType.SEMICOLON);
                break;
            case '*':
                addToken(TokenType.STAR);
                break;
            case '%':
                addToken(TokenType.MODULO);
                break;
            case '!':
                addToken(match('=') ? TokenType.BANG_EQUAL : TokenType.BANG);
                break;
            case '=':
                addToken(match('=') ? TokenType.EQUAL_EQUAL : TokenType.EQUAL);
                break;
            case '<':
                addToken(match('=') ? TokenType.LESS_EQUAL : TokenType.LESS);
                break;
            case '>':
                addToken(match('=') ? TokenType.GREATER_EQUAL : TokenType.GREATER);
                break;
            case '/':
                if (match('/')) {
                    // A comment goes until the end of the line.
                    while (peek() != '\n' && !isAtEnd())
                        advance();
                    // addToken(TokenType.COMMENT);
                } else if (match('*')) {
                    // A block comment goes until '*/' termination
                    while (!isAtEnd() && (peek() != '*' && peekNext() != '/'))
                        advance();
                    // consume */
                    advance(2);
                    // addToken(TokenType.COMMENT);

                } else {
                    addToken(TokenType.SLASH);
                }
                break;
            // whitespace
            case ' ':
            case '\r':
            case '\t':
                break;
            case '\n':
                line++;
                break;
            // string literals
            case '"':
                string();
                break;
            default:
                if (isDigit(c)) {
                    number();
                } else if (isAlpha(c)) {
                    identifier();
                } else {
                    Grommet.error(line, "Unexpected character.");
                }
                break;
        }
    }

    private void identifier() {
        while (isAlphaNumeric(peek()))
            advance();

        String text = source.substring(start, current);
        TokenType type = keywords.get(text);
        if (type == null)
            type = TokenType.IDENTIFIER;

        addToken(type);
    }

    private boolean isAlpha(char c) {
        return (c >= 'a' && c <= 'z') ||
                (c >= 'A' && c <= 'Z') || (c == '_');
    }

    private boolean isAlphaNumeric(char c) {
        return isAlpha(c) || isDigit(c);
    }

    private void number() {
        while (isDigit(peek()))
            advance();

        // look for a fractional part
        if (peek() == '.' && isDigit(peekNext()))
            advance(); // consume the .

        while (isDigit(peek()))
            advance();

        double value = Double.parseDouble(source.substring(start, current));
        addToken(TokenType.NUMBER, value);
    }

    private void string() {
        while (peek() != '"' && !isAtEnd()) {
            if (peek() == '\n') // lox supports multi-line strings
                line++;

            advance();
        }

        if (isAtEnd()) {
            Grommet.error(line, "Unterminiated string.");
            return;
        }
        // the closing "
        advance();

        // Trim the surrounding quotes
        String value = source.substring(start + 1, current - 1);
        addToken(TokenType.STRING, value);
    }

    private boolean match(char expected) {
        if (isAtEnd())
            return false;
        if (this.source.charAt(current) != expected)
            return false;

        this.current++;
        return true;
    }

    private char peek() {
        if (isAtEnd())
            return '\0';
        return this.source.charAt(current);
    }

    private char peekNext() {
        if (current + 1 >= source.length())
            return '\0';
        return source.charAt(current + 1);
    }

    private boolean isDigit(char c) {
        return c >= '0' && c <= '9';
    }

    private boolean isAtEnd() {
        return current >= source.length();
    }

    // consumes th next char in the source
    private char advance() {
        return source.charAt(current++);
    }

    private void advance(int n) {
        for (int i = 0; i < n; i++) {
            advance();
        }
    }

    // creates a new token for the curernt lexeme
    private void addToken(TokenType type) {
        addToken(type, null);
    }

    private void addToken(TokenType type, Object literal) {
        String text = source.substring(start, current);
        tokens.add(new Token(type, text, literal, line));
    }
}

File: src/scanner.go
package main

import (
	"fmt"
	"gourmet/reserved"
	"gourmet/tokens"
	"gourmet/util"
	"strconv"
	"strings"
	"unicode"
	"unicode/utf8"
)

var (
	src     string
	srcLen  int
	tkns    = []tokens.Token{}
	start   int
	current int = 0
	line    int = 1
)

func ScanTokens(input string) {
	src = input
	srcLen = utf8.RuneCountInString(src)
	for !isAtEnd() {
		// start at the beginning of the next lexeme
		start = current
		scanToken()
	}
	for _, v := range tkns {
		fmt.Println(v)
	}
}

func scanToken() {
	r := advance()
	switch r {
	case '(':
		addToken(tokens.LEFT_PAREN)
	case ')':
		addToken(tokens.RIGHT_PAREN)
	case '{':
		addToken(tokens.LEFT_BRACE)
	case '}':
		addToken(tokens.RIGHT_BRACE)
	case ',':
		addToken(tokens.COMMA)
	case '.':
		addToken(tokens.DOT)
	case '-':
		if match('=') {
			addToken(tokens.MINUSEQUALS)
		} else {
			addToken(tokens.MINUS)
		}
	case '+':
		if match('=') {
			addToken(tokens.PLUSEQUALS)
		} else {
			addToken(tokens.PLUS)
		}
	case ';':
		addToken(tokens.SEMICOLON)
	case '*':
		addToken(tokens.STAR)
	case '%':
		addToken(tokens.MODULO)
	case '!':
		if match('=') {
			addToken(tokens.BANG_EQUAL)
		} else {
			addToken(tokens.BANG)
		}
	case ':':
		addToken(tokens.COLON)
	case '=':
		if match('=') {
			addToken(tokens.EQUAL_EQUAL)
		} else {
			addToken(tokens.EQUAL)
		}
	case '<':
		if match('=') {
			addToken(tokens.LESS_EQUAL)
		} else {
			addToken(tokens.LESS)
		}
	case '>':
		if match('=') {
			addToken(tokens.GREATER_EQUAL)
		} else {
			addToken(tokens.GREATER)
		}
	case '/':
		comment()
	// whitespace
	case ' ', '\r', '\t': // do nothing
	case '\n':
		line++
	// string literals
	case '"':
		str()
	default:
		if isDigit(r) {
			number()
		} else if isAlpha(r) || r == '@' {
			identifier()
		} else {
			fmt.Println("--->", string(r))
			fmt.Println(line)
			panic("Unexpected character.")
			// Grommet.error(line, "Unexpected character.")
		}

	}
}

// Check if the current rune equal to expected
func match(expected rune) bool {
	if isAtEnd() {
		return false
	}
	r, _ := utf8.DecodeRuneInString(src[current:])

	if r != expected {
		return false
	}

	current++
	return true
}

// peek the current rune without consuming
func peek() rune {
	if isAtEnd() {
		return '•' // this means eof
	}
	r, _ := utf8.DecodeRuneInString(src[current:])

	return r
}

// peek the next rune without consuming
func peekNext() rune {
	if current+1 >= srcLen {
		return '•'
	}
	r, _ := utf8.DecodeRuneInString(src[current+1:])
	return r
}

// handle comments
func comment() {
	if match('/') {
		// A comment goes until the end of the line.
		for peek() != '\n' && !isAtEnd() {
			advance()
		}
		addToken(tokens.COMMENT)
	} else if match('*') {
		// A block comment goes until '*/' termination
		for !isAtEnd() && (peek() != '*' && peekNext() != '/') {
			if peek() == '\n' { // supports multi-line comments
				line++
			}
			advance()
		}
		// consume */
		advanceN(2)
		addToken(tokens.BLOCK_COMMENT)

	} else {
		addToken(tokens.SLASH)
	}
}

// handle strings
func str() {
	for peek() != '"' && !isAtEnd() {
		if peek() == '\n' { // supports multi-line strings
			line++
		}

		advance()
	}

	if isAtEnd() {
		panic("Unterminiated string.")
		// Grommet.error(line, "Unterminiated string.")
		// return
	}
	// the closing "
	advance()

	// Trim the surrounding quotes
	value := src[start+1 : current-1]
	addTokenLit(tokens.STRING, value)
}

// check if the rune is a digit
func isDigit(r rune) bool {
	return r >= '0' && r <= '9'
}

// handle numbers
func number() {
	for isDigit(peek()) {
		advance()
	}

	// look for a fractional part
	if peek() == '.' && isDigit(peekNext()) {
		advance()
	} // consume the .

	for isDigit(peek()) {
		advance()
	}

	value, err := strconv.ParseFloat(src[start:current], 32)
	util.Check(err)
	addTokenLit(tokens.NUMBER, value)
}

// check if the rune is an alpha char
func isAlpha(r rune) bool {
	return (r >= 'a' && r <= 'z') ||
		(r >= 'A' && r <= 'Z') || (r == '_')
}
func isAlphaNumeric(r rune) bool {
	return isAlpha(r) || isDigit(r)
}

// handle identifiers
func identifier() {
	for isAlphaNumeric(peek()) || peek() == '@' {
		advance()
	}

	text := strings.ToLower(src[start:current])
	ttype, exists := reserved.Keywords[text]
	if !exists {
		ttype, exists = reserved.Types[text]
	}
	if !exists {
		ttype = tokens.IDENTIFIER
	}

	// check for typdef tuple
	if text == "typedef" {
		advance()
		prev := current
		nextToken()
		w := strings.ToLower(src[prev:current])
		w = strings.ReplaceAll(w, " ", "")
		w = strings.ReplaceAll(w, "\t", "")
		if w == "tuple" {
			ttype = tokens.TYPEDEF
		}
	}
	addToken(ttype)
}

// get the next token (skipping whitespace)
func nextToken() {
	// find the next word
	for unicode.IsSpace(peek()) {
		advance()
	}
	// find the end of the word

	for !unicode.IsSpace(peek()) {
		advance()
	}
}

// creates a new token for the curernt lexeme
func addToken(ttype tokens.TokenType) {
	addTokenLit(ttype, nil)
}

func addTokenLit(ttype tokens.TokenType, literal any) {
	text := src[start:current]
	tkns = append(tkns, tokens.Token{TokenType: ttype, Lexeme: text, Literal: literal, Line: line})
}

// consumes th next char in the source
func advance() rune {
	r, w := utf8.DecodeRuneInString(src[current:])
	current += w
	return r

}
func advanceN(n int) {
	for i := 0; i < n; i++ {
		advance()
	}
}
func isAtEnd() bool {
	return current >= srcLen
}
